ystia/yorc
# Feature/async job run

# Pull Request description
This PR implements #93
This PR fixes #161
## Description of the change
### What I did
Dissociate the job launching and its monitoring in order to make the run step job asynchronous and not to block a specific worker per job. The run workflow step is in “running” status until the monitoring detects it’s finished and updates its status to “done” then resumes the workflow.

### How I did it
- Define **tosca.interfaces.node.lifecycle.runnable.run** as asynchronous operation.
- Add new function **ExecAsyncOperation** for **OperationExecutor** interface. This allow to return an acton and a time interval to schedule its monitoring.
- Implement this function for slurm operation executor and disable the ExecOperation in order to handle only async operations for slurm jobs

**Implement an **Action** and **scheduled executions** system**
- Add **Action** Task-typed management
- Action is defined with **actionType** to specify the type of execution and **data** to provide useful information for action execution.
- Create **ActionOperator** as the component responsible for executing action via TaskManager
- Add **ActionOperator** in **registry** with a match on **actionType**
- Create a **scheduler** to handle scheduled actions (simple Action with defined time Interval to execute it periodically). Handle it in a resilient way with Consul leader election.
- Implement a proceed function for **scheduledAction** which allows to register a task action-typed in order to let it be executed by Task workers.
- Allow the possibility to register/unregister a **scheduledAction** in the scheduling API

**Use the scheduler and the actionOperator to monitor a job**
- Register the scheduled “job-monitoring” action type when **tosca.interfaces.node.lifecycle.runnable.run** operation is executed
- Allow the possibility to configure the job monitoring time interval in slurm infrastructure with **job_monitoring_time_interval property** or in Tosca Slurm job property with **monitoring_time_interval**. Otherwise, default time interval is 5s

### How to verify it
Deploy https://github.com/ystia/yorc-a4c-plugin/tree/develop/tosca-samples/org/ystia/yorc/samples/job/sbatch
https://github.com/ystia/yorc-a4c-plugin/tree/develop/tosca-samples/org/ystia/yorc/samples/job/srun

See the logs and watch periodic status changes.
Check the job monitoring is stopped and the workflow is done when the job ends
